{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import pathlib\n",
    "import pysam\n",
    "import pyfaidx\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "output_version = \"dev\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All output are saved at /media/hieunguyen/HNSD_mini/outdir/targetMethyl_analysis/dev/Vi_output/non_directional\n"
     ]
    }
   ],
   "source": [
    "maindir = \"/media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis\"\n",
    "outputdir = \"/media/hieunguyen/HNSD_mini/outdir\"\n",
    "\n",
    "Vi_runs = [\"R6801\", \"R6829Vi\", \"R6873\", \"R6914\"]\n",
    "Truong_runs = [\"R6782\", \"R6823\", \"R6829Truong\"]\n",
    "\n",
    "regions = {\n",
    "    \"Vi\": \"Vi_amplicons.hg19.bed\",\n",
    "    \"Truong\": \"Truong_amplicons.hg19.bed\"\n",
    "}\n",
    "\n",
    "regions_hg38 = {\n",
    "    \"Vi\": \"Vi_amplicons.hg38.bed\",\n",
    "    \"Truong\": \"Truong_amplicons.hg38.bed\"\n",
    "}\n",
    "\n",
    "# mode = \"directional\"\n",
    "mode = \"non_directional\"\n",
    "\n",
    "# pic = \"Truong\"\n",
    "pic = \"Vi\"\n",
    "\n",
    "path_to_main_output = os.path.join(outputdir, \"targetMethyl_analysis\", output_version, f\"{pic}_output\", mode)\n",
    "os.system(f\"mkdir -p {path_to_main_output}\")\n",
    "\n",
    "print(f\"All output are saved at {path_to_main_output}\")\n",
    "\n",
    "\n",
    "real_cpgdf = pd.read_excel(os.path.join(path_to_main_output, \"00_output\", f\"{pic}_panel_correct_cpgdf.xlsx\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>Run</th>\n",
       "      <th>path</th>\n",
       "      <th>mode</th>\n",
       "      <th>PIC</th>\n",
       "      <th>bam_path</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TM010</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TM100</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TM002</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TM050</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TM005</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TM001</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3-TMH3_S7527-S7727</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>4-TMH4_S7528-S7728</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>11-TMM3_S7535-S7735</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1-TMH1_S7525-S7725</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10-TMM2_S7534-S7734</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>6-TMS1_S7530-S7730</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>7-TMS2_S7531-S7731</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>2-TMH2_S7526-S7726</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>9-TMM1_S7533-S7733</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>8-TMS3_S7532-S7732</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>5-TMH5_S7529-S7729</td>\n",
       "      <td>R6873</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>7-TMKA71A1_S7531-S7731</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>9-TMLBE329_S7533-S7733</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>5-TMKA67A1_S7529-S7729</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1-TMKC25A1_S7525-S7725</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>8-TMKA72A1_S7532-S7732</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>4-TMKC53A1_S7528-S7728</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>10-TMKE30A1_S7534-S7734</td>\n",
       "      <td>R6829Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>TMM3P1</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>TMM3P3</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>TMM3P2</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>TMS2</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TMS6</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TMM3P4</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>TMM1P4</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>TMS1</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>TMS3</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>TMM1P1</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>TMS4</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>TMM1P2</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>TMM2P3</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>TMM2P2</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>TMS5</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>TMM2P4</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>TMM2P1</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>TMM1P3</td>\n",
       "      <td>R6914</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   filename      Run  \\\n",
       "0                     TM010    R6801   \n",
       "1                     TM100    R6801   \n",
       "2                     TM002    R6801   \n",
       "3                     TM050    R6801   \n",
       "4                     TM005    R6801   \n",
       "5                     TM001    R6801   \n",
       "29       3-TMH3_S7527-S7727    R6873   \n",
       "30       4-TMH4_S7528-S7728    R6873   \n",
       "31      11-TMM3_S7535-S7735    R6873   \n",
       "32       1-TMH1_S7525-S7725    R6873   \n",
       "33      10-TMM2_S7534-S7734    R6873   \n",
       "34       6-TMS1_S7530-S7730    R6873   \n",
       "35       7-TMS2_S7531-S7731    R6873   \n",
       "36       2-TMH2_S7526-S7726    R6873   \n",
       "37       9-TMM1_S7533-S7733    R6873   \n",
       "38       8-TMS3_S7532-S7732    R6873   \n",
       "39       5-TMH5_S7529-S7729    R6873   \n",
       "40   7-TMKA71A1_S7531-S7731  R6829Vi   \n",
       "41   9-TMLBE329_S7533-S7733  R6829Vi   \n",
       "42   5-TMKA67A1_S7529-S7729  R6829Vi   \n",
       "43   1-TMKC25A1_S7525-S7725  R6829Vi   \n",
       "44   8-TMKA72A1_S7532-S7732  R6829Vi   \n",
       "45   4-TMKC53A1_S7528-S7728  R6829Vi   \n",
       "46  10-TMKE30A1_S7534-S7734  R6829Vi   \n",
       "59                   TMM3P1    R6914   \n",
       "60                   TMM3P3    R6914   \n",
       "61                   TMM3P2    R6914   \n",
       "62                     TMS2    R6914   \n",
       "63                     TMS6    R6914   \n",
       "64                   TMM3P4    R6914   \n",
       "65                   TMM1P4    R6914   \n",
       "66                     TMS1    R6914   \n",
       "67                     TMS3    R6914   \n",
       "68                   TMM1P1    R6914   \n",
       "69                     TMS4    R6914   \n",
       "70                   TMM1P2    R6914   \n",
       "71                   TMM2P3    R6914   \n",
       "72                   TMM2P2    R6914   \n",
       "73                     TMS5    R6914   \n",
       "74                   TMM2P4    R6914   \n",
       "75                   TMM2P1    R6914   \n",
       "76                   TMM1P3    R6914   \n",
       "\n",
       "                                                 path             mode PIC  \\\n",
       "0   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "1   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "2   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "3   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "4   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "5   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "29  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "30  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "31  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "32  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "33  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "34  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "35  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "36  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "37  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "38  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "39  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "40  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "41  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "42  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "43  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "44  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "45  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "46  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "59  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "60  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "61  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "62  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "63  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "64  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "65  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "66  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "67  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "68  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "69  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "70  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "71  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "72  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "73  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "74  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "75  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "76  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  non_directional  Vi   \n",
       "\n",
       "                                             bam_path  \n",
       "0   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "1   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "2   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "3   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "4   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "5   /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "29  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "30  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "31  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "32  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "33  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "34  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "35  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "36  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "37  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "38  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "39  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "40  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "41  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "42  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "43  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "44  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "45  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "46  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "59  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "60  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "61  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "62  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "63  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "64  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "65  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "66  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "67  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "68  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "69  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "70  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "71  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "72  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "73  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "74  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "75  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  \n",
       "76  /media/hieunguyen/HNHD01/raw_data/targetMethyl...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_cov_files = [item for item in pathlib.Path(maindir).glob(\"target_methylation_*/06*/*.cov\")]\n",
    "\n",
    "metadata_dict = {\n",
    "    \"filename\"  : [item.name.split(\".no_deduplicated\")[0] for item in all_cov_files],\n",
    "    \"Run\" : [str(item).split(\"/\")[6].replace(\"_no_dedup\", \"\").replace(\"target_methylation_\", \"\") for item in all_cov_files],\n",
    "    \"path\": [str(item) for item in all_cov_files]\n",
    "}\n",
    "\n",
    "metadata = pd.DataFrame.from_dict(metadata_dict, orient=\"columns\")\n",
    "metadata[\"mode\"] = metadata[\"Run\"].apply(lambda x: \"directional\" if \"_without_non_directional\" in x else \"non_directional\")\n",
    "metadata[\"Run\"] = metadata[\"Run\"].apply(lambda x: x.split(\"_\")[0])\n",
    "metadata[\"PIC\"] = metadata[\"Run\"].apply(lambda x: \"Vi\" if x in Vi_runs else \"Truong\")\n",
    "metadata[\"bam_path\"] = metadata[\"path\"].apply(lambda x: str(x).replace(\"06_methylation_extract\", \"05_sorted_bam\").replace(\".bedGraph.gz.bismark.zero.cov\", \".sorted.bam\"))\n",
    "metadata = metadata[(metadata[\"mode\"] == mode) & (metadata[\"PIC\"] == pic)]\n",
    "\n",
    "##### regiondf for hg19\n",
    "regiondf = pd.read_csv(regions[pic], sep = \"\\t\", header = None)\n",
    "regiondf.columns = [\"chrom\", \"start\", \"end\"]\n",
    "regiondf = regiondf[[\"chrom\", \"start\", \"end\"]]\n",
    "regiondf[\"region_name\"] = regiondf[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"region_{x[0]}_{x[1]}_{x[2]}\", axis = 1\n",
    ")\n",
    "regiondf[\"bam_region\"] = regiondf[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"{x[0].replace('chr', '')}:{x[1]}-{x[2]}\", axis = 1\n",
    ")\n",
    "\n",
    "##### regiondf for hg38\n",
    "regiondf_hg38 = pd.read_csv(regions_hg38[pic], sep = \",\", header = None)\n",
    "regiondf_hg38.columns = [\"chrom\", \"start\", \"end\"]\n",
    "regiondf_hg38 = regiondf_hg38[[\"chrom\", \"start\", \"end\"]]\n",
    "regiondf_hg38[\"region_name\"] = regiondf_hg38[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"region_{x[0]}_{x[1]}_{x[2]}\", axis = 1\n",
    ")\n",
    "regiondf_hg38[\"bam_region\"] = regiondf_hg38[[\"chrom\", \"start\", \"end\"]].apply(\n",
    "    lambda x: f\"{x[0].replace('chr', '')}:{x[1]}-{x[2]}\", axis = 1\n",
    ")\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Count on/off target reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists, skip counting reads in regions. Reading existing data... \n"
     ]
    }
   ],
   "source": [
    "metadata[\"num_total_reads\"] = metadata[\"bam_path\"].apply(lambda x: int(pysam.samtools.view(\"-c\", x, catch_stdout=True)))\n",
    "\n",
    "def count_read_in_region(bam_path, region, chr_mode = False):\n",
    "    all_reads = []\n",
    "    bamfile = pysam.AlignmentFile(bam_path, \"rb\")\n",
    "    if chr_mode:\n",
    "        region = f\"chr{region}\"\n",
    "    fetched_obj = bamfile.fetch(region = region)\n",
    "    for read in fetched_obj:\n",
    "        all_reads.append(read)\n",
    "    return(len(all_reads))\n",
    "if os.path.isfile(os.path.join(path_to_main_output, f\"{pic}_read_count_in_region.xlsx\")) == False:\n",
    "    for region_name in regiondf.region_name.unique():\n",
    "        bam_region = regiondf[regiondf[\"region_name\"] == region_name][\"bam_region\"].values[0]\n",
    "        metadata[region_name] = metadata[\"bam_path\"].apply(lambda x: count_read_in_region(x, bam_region))\n",
    "        metadata[f\"pct_{region_name}\"] = metadata[region_name] / metadata[\"num_total_reads\"] * 100\n",
    "        \n",
    "    metadata.to_excel(os.path.join(path_to_main_output, f\"{pic}_read_count_in_region.xlsx\"), index = False)\n",
    "else:\n",
    "    print(\"File already exists, skip counting reads in regions. Reading existing data... \")\n",
    "    metadata = pd.read_excel(os.path.join(path_to_main_output, f\"{pic}_read_count_in_region.xlsx\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process cov file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>Run</th>\n",
       "      <th>path</th>\n",
       "      <th>mode</th>\n",
       "      <th>PIC</th>\n",
       "      <th>bam_path</th>\n",
       "      <th>num_total_reads</th>\n",
       "      <th>region_chr16_22825587_22825729</th>\n",
       "      <th>pct_region_chr16_22825587_22825729</th>\n",
       "      <th>region_chr16_22825770_22825948</th>\n",
       "      <th>...</th>\n",
       "      <th>region_chr16_22826024_22826152</th>\n",
       "      <th>pct_region_chr16_22826024_22826152</th>\n",
       "      <th>region_chr16_22826130_22826286</th>\n",
       "      <th>pct_region_chr16_22826130_22826286</th>\n",
       "      <th>region_chr17_75369424_75369590</th>\n",
       "      <th>pct_region_chr17_75369424_75369590</th>\n",
       "      <th>region_chr17_75369739_75369887</th>\n",
       "      <th>pct_region_chr17_75369739_75369887</th>\n",
       "      <th>region_chr17_75369604_75369763</th>\n",
       "      <th>pct_region_chr17_75369604_75369763</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TM010</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TM100</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>36</td>\n",
       "      <td>16</td>\n",
       "      <td>44.444444</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TM002</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TM050</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>38</td>\n",
       "      <td>14</td>\n",
       "      <td>36.842105</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TM005</td>\n",
       "      <td>R6801</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>non_directional</td>\n",
       "      <td>Vi</td>\n",
       "      <td>/media/hieunguyen/HNHD01/raw_data/targetMethyl...</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  filename    Run                                               path  \\\n",
       "0    TM010  R6801  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "1    TM100  R6801  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "2    TM002  R6801  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "3    TM050  R6801  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "4    TM005  R6801  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "\n",
       "              mode PIC                                           bam_path  \\\n",
       "0  non_directional  Vi  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "1  non_directional  Vi  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "2  non_directional  Vi  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "3  non_directional  Vi  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "4  non_directional  Vi  /media/hieunguyen/HNHD01/raw_data/targetMethyl...   \n",
       "\n",
       "   num_total_reads  region_chr16_22825587_22825729  \\\n",
       "0               26                               0   \n",
       "1               36                              16   \n",
       "2               32                               0   \n",
       "3               38                              14   \n",
       "4               10                               0   \n",
       "\n",
       "   pct_region_chr16_22825587_22825729  region_chr16_22825770_22825948  ...  \\\n",
       "0                            0.000000                               0  ...   \n",
       "1                           44.444444                               0  ...   \n",
       "2                            0.000000                               0  ...   \n",
       "3                           36.842105                               0  ...   \n",
       "4                            0.000000                               0  ...   \n",
       "\n",
       "   region_chr16_22826024_22826152  pct_region_chr16_22826024_22826152  \\\n",
       "0                               0                                 0.0   \n",
       "1                               0                                 0.0   \n",
       "2                               0                                 0.0   \n",
       "3                               0                                 0.0   \n",
       "4                               0                                 0.0   \n",
       "\n",
       "   region_chr16_22826130_22826286  pct_region_chr16_22826130_22826286  \\\n",
       "0                               0                                 0.0   \n",
       "1                               0                                 0.0   \n",
       "2                               0                                 0.0   \n",
       "3                               0                                 0.0   \n",
       "4                               0                                 0.0   \n",
       "\n",
       "   region_chr17_75369424_75369590  pct_region_chr17_75369424_75369590  \\\n",
       "0                               0                                 0.0   \n",
       "1                               0                                 0.0   \n",
       "2                               0                                 0.0   \n",
       "3                               0                                 0.0   \n",
       "4                               0                                 0.0   \n",
       "\n",
       "   region_chr17_75369739_75369887  pct_region_chr17_75369739_75369887  \\\n",
       "0                               0                                 0.0   \n",
       "1                               0                                 0.0   \n",
       "2                               0                                 0.0   \n",
       "3                               0                                 0.0   \n",
       "4                               0                                 0.0   \n",
       "\n",
       "   region_chr17_75369604_75369763  pct_region_chr17_75369604_75369763  \n",
       "0                               0                                 0.0  \n",
       "1                               0                                 0.0  \n",
       "2                               0                                 0.0  \n",
       "3                               0                                 0.0  \n",
       "4                               0                                 0.0  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>region_name</th>\n",
       "      <th>bam_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr16</td>\n",
       "      <td>22825587</td>\n",
       "      <td>22825729</td>\n",
       "      <td>region_chr16_22825587_22825729</td>\n",
       "      <td>16:22825587-22825729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>22825770</td>\n",
       "      <td>22825948</td>\n",
       "      <td>region_chr16_22825770_22825948</td>\n",
       "      <td>16:22825770-22825948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr16</td>\n",
       "      <td>22825920</td>\n",
       "      <td>22826053</td>\n",
       "      <td>region_chr16_22825920_22826053</td>\n",
       "      <td>16:22825920-22826053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>22826024</td>\n",
       "      <td>22826152</td>\n",
       "      <td>region_chr16_22826024_22826152</td>\n",
       "      <td>16:22826024-22826152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16</td>\n",
       "      <td>22826130</td>\n",
       "      <td>22826286</td>\n",
       "      <td>region_chr16_22826130_22826286</td>\n",
       "      <td>16:22826130-22826286</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end                     region_name  \\\n",
       "0  chr16  22825587  22825729  region_chr16_22825587_22825729   \n",
       "1  chr16  22825770  22825948  region_chr16_22825770_22825948   \n",
       "2  chr16  22825920  22826053  region_chr16_22825920_22826053   \n",
       "3  chr16  22826024  22826152  region_chr16_22826024_22826152   \n",
       "4  chr16  22826130  22826286  region_chr16_22826130_22826286   \n",
       "\n",
       "             bam_region  \n",
       "0  16:22825587-22825729  \n",
       "1  16:22825770-22825948  \n",
       "2  16:22825920-22826053  \n",
       "3  16:22826024-22826152  \n",
       "4  16:22826130-22826286  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regiondf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_covdf = dict()\n",
    "for run in metadata.Run.unique():\n",
    "      for filename in metadata[metadata[\"Run\"] == run][\"filename\"].unique():\n",
    "            path_to_save_cov = os.path.join(path_to_main_output, run, filename)\n",
    "            os.system(f\"mkdir -p {path_to_save_cov}\")\n",
    "\n",
    "            covdf = pd.read_csv(metadata[metadata[\"filename\"] == filename][\"path\"].values[0], header = None, sep = \"\\t\")\n",
    "            covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "            all_covdf[filename] = covdf\n",
    "            covdf[\"chrom\"] = covdf[\"chrom\"].apply(lambda x: str(x))\n",
    "            for region in regiondf.region_name.unique():\n",
    "                  chrom = str(region.split(\"_\")[1].replace(\"chr\", \"\"))\n",
    "                  start = int(region.split(\"_\")[2])\n",
    "                  end = int(region.split(\"_\")[3])\n",
    "                  save_covdf = covdf[(covdf[\"chrom\"] == chrom) & \n",
    "                        (covdf[\"start\"] >= start) & (covdf[\"start\"] <= end)]\n",
    "                  save_covdf[\"CpG\"] = save_covdf[[\"chrom\", \"start\", \"end\"]].apply(lambda x: f\"chr{str(x[0])}:{x[1]}-{x[2]}\", axis = 1)\n",
    "                  save_covdf[\"check_context\"] = save_covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in real_cpgdf[\"CpG\"].values else \"False\")\n",
    "                  save_covdf.to_excel(f\"{path_to_save_cov}/{region}.xlsx\", index = False)\n",
    "                  # print(f\"Sample {filename}, Run {run}, region {region}, number of reads = {save_covdf.shape}\")\n",
    "                  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>meth_density</th>\n",
       "      <th>countC</th>\n",
       "      <th>countT</th>\n",
       "      <th>CpG</th>\n",
       "      <th>check_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, meth_density, countC, countT, CpG, check_context]\n",
       "Index: []"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "save_covdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check sample: CONTROL114, R6447"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check number of reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16, 6)\n",
      "(19, 6)\n",
      "(11, 6)\n",
      "(14, 6)\n",
      "(10, 6)\n",
      "(12, 6)\n",
      "(17, 6)\n",
      "(14, 6)\n",
      "(16, 6)\n",
      "(19, 6)\n",
      "(11, 6)\n",
      "(14, 6)\n",
      "(10, 6)\n",
      "(12, 6)\n",
      "(17, 6)\n",
      "(14, 6)\n"
     ]
    }
   ],
   "source": [
    "control114_cov = \"/media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.cg_cov.bed\"\n",
    "control114_bam = \"/media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam\"\n",
    "control114_cov_hg19 = \"/media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.cg_cov.hg19.bed\"\n",
    "\n",
    "\n",
    "path_to_save_cov = os.path.join(path_to_main_output, \"R6447\", \"CONTROL114\")\n",
    "os.system(f\"mkdir -p {path_to_save_cov}\")\n",
    "os.system(f\"mkdir -p {path_to_save_cov}/hg38\")\n",
    "os.system(f\"mkdir -p {path_to_save_cov}/hg19\")\n",
    "\n",
    "#### data target commercial 1-based, our data is 0-based\n",
    "covdf = pd.read_csv(control114_cov, header = None, sep = \"\\t\")\n",
    "covdf.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "covdf[\"start\"] = covdf[\"start\"].apply(lambda x: int(x) - 1)\n",
    "covdf[\"end\"] = covdf[\"end\"].apply(lambda x: int(x) - 1)\n",
    "\n",
    "covdfhg19 = pd.read_csv(control114_cov_hg19, header = None, sep = \"\\t\")\n",
    "covdfhg19.columns = [\"chrom\", \"start\", \"end\", \"meth_density\", \"countC\", \"countT\"]\n",
    "covdfhg19[\"start\"] = covdfhg19[\"start\"].apply(lambda x: int(x) - 1)\n",
    "covdfhg19[\"end\"] = covdfhg19[\"end\"].apply(lambda x: int(x) - 1)\n",
    "\n",
    "\n",
    "covdfhg19[\"countC\"] = covdf[\"countC\"].values\n",
    "covdfhg19[\"countT\"] = covdf[\"countT\"].values\n",
    "\n",
    "for region in regiondf_hg38.region_name.unique():\n",
    "        chrom = region.split(\"_\")[1]\n",
    "        start = int(region.split(\"_\")[2])\n",
    "        end = int(region.split(\"_\")[3])\n",
    "\n",
    "        covdf[(covdf[\"chrom\"] == chrom) & \n",
    "            (covdf[\"start\"] >= start) & (covdf[\"start\"] <= end)].to_excel(f\"{path_to_save_cov}/hg38/{region}.xlsx\", index = False)\n",
    "        \n",
    "        print(covdf[(covdf[\"chrom\"] == chrom) & \n",
    "            (covdf[\"start\"] >= start) & (covdf[\"start\"] <= end)].shape)\n",
    "        \n",
    "for region in regiondf.region_name.unique():\n",
    "        chrom = region.split(\"_\")[1]\n",
    "        start = int(region.split(\"_\")[2])\n",
    "        end = int(region.split(\"_\")[3])\n",
    "        save_covdf = covdfhg19[(covdfhg19[\"chrom\"] == chrom) & \n",
    "            (covdfhg19[\"start\"] >= start) & (covdfhg19[\"start\"] <= end)]\n",
    "        save_covdf[\"CpG\"] = save_covdf[[\"chrom\", \"start\", \"end\"]].apply(lambda x: f\"{str(x[0])}:{x[1]}-{x[2]}\", axis = 1)\n",
    "        save_covdf[\"check_context\"] = save_covdf[\"CpG\"].apply(lambda x: \"CpG_context\" if x in real_cpgdf[\"CpG\"].values else \"False\")\n",
    "                  \n",
    "        save_covdf.to_excel(f\"{path_to_save_cov}/hg19/{region}.xlsx\", index = False)\n",
    "        \n",
    "        print(covdfhg19[(covdfhg19[\"chrom\"] == chrom) & \n",
    "            (covdfhg19[\"start\"] >= start) & (covdfhg19[\"start\"] <= end)].shape)\n",
    "        \n",
    "all_covdf[\"CONTROL114\"] = covdfhg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n",
      "[W::hts_idx_load3] The index file is older than the data file: /media/hieunguyen/HNHD01/raw_data/targetMethyl_analysis/CONTROL114/65-CONTROL114CT836T_M550-M750.deduplicated.sort.bam.bai\n"
     ]
    }
   ],
   "source": [
    "metadata_CONTROL114 = pd.DataFrame.from_dict({\n",
    "    \"filename\": [\"CONTROL114\"],\n",
    "    \"Run\": [\"R6447\"],\n",
    "    \"bam_path\": [control114_bam],\n",
    "    \"path\": [control114_cov]\n",
    "})\n",
    "\n",
    "metadata_CONTROL114[\"num_total_reads\"] = metadata_CONTROL114[\"bam_path\"].apply(lambda x: int(pysam.samtools.view(\"-c\", x, catch_stdout=True)))\n",
    "\n",
    "for region_name in regiondf_hg38.region_name.unique():\n",
    "    bam_region = regiondf_hg38[regiondf_hg38[\"region_name\"] == region_name][\"bam_region\"].values[0]\n",
    "    metadata_CONTROL114[region_name] = metadata_CONTROL114[\"bam_path\"].apply(lambda x: count_read_in_region(x, bam_region, chr_mode = True))\n",
    "    metadata_CONTROL114[f\"pct_{region_name}\"] = metadata_CONTROL114[region_name] / metadata_CONTROL114[\"num_total_reads\"] * 100\n",
    "\n",
    "metadata_CONTROL114.to_excel(os.path.join(path_to_main_output, f\"CONTROL114_read_count_in_region.xlsx\"), index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare methylation density between 2 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_methyldf = pd.DataFrame(data = regiondf.region_name.unique(), columns = [\"region\"])\n",
    "\n",
    "def get_mean_methyl_in_region(region, filename, remove_non_cpg = True):\n",
    "\n",
    "    if filename == \"CONTROL114\":\n",
    "        df = pd.read_excel(f\"{os.path.join(path_to_main_output, 'R6447', 'CONTROL114')}/hg19/{region}.xlsx\")\n",
    "        run = \"R6447\"\n",
    "    else:\n",
    "        run = metadata[metadata[\"filename\"] == filename][\"Run\"].values[0]\n",
    "        df = pd.read_excel(f\"{os.path.join(path_to_main_output, run, filename)}/{region}.xlsx\")\n",
    "    if df.shape[0] == 0:\n",
    "        return \"no data available\"\n",
    "    else:\n",
    "        if remove_non_cpg:\n",
    "            # keep only CpG context Cs\n",
    "            df = df[df[\"check_context\"] == \"CpG_context\"]\n",
    "        mean_methyl = df.meth_density.mean()\n",
    "        return mean_methyl\n",
    "\n",
    "for filename in all_covdf.keys():\n",
    "    mean_methyldf[filename] = mean_methyldf[\"region\"].apply(lambda x: get_mean_methyl_in_region(x, filename))\n",
    "\n",
    "mean_methyldf.to_excel(os.path.join(path_to_main_output, f\"{pic}_mean_methyl_in_region_compare_CONTROL114.xlsx\"), index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_double_checkdf = pd.DataFrame(data = regiondf.region_name.unique(), columns = [\"region\"])\n",
    "for run in metadata.Run.unique():\n",
    "      for filename in metadata[metadata[\"Run\"] == run][\"filename\"].unique():\n",
    "            path_to_save_double_check_file = os.path.join(path_to_main_output, run, filename, \"check_MethylDensity_and_Count\")\n",
    "            os.system(f\"mkdir -p {path_to_save_double_check_file}\")\n",
    "            check_region_countdf = metadata[metadata[\"filename\"] == filename]\\\n",
    "                [[item for item in metadata.columns if \"region\" in item and \"pct\" not in item]]\\\n",
    "                    .T.reset_index()\n",
    "            check_region_countdf.columns = [\"region\", f\"count_{filename}\"]\n",
    "            check_methyldf = mean_methyldf[[\"region\", filename]]\n",
    "            double_checkdf = check_methyldf.merge(check_region_countdf, right_on = \"region\", left_on = \"region\")\n",
    "            double_checkdf.to_excel(f\"{path_to_save_double_check_file}/check_methylDensity_and_Count_{filename}.xlsx\", index = False)\n",
    "            merge_double_checkdf = merge_double_checkdf.merge(double_checkdf, right_on = \"region\", left_on = \"region\")\n",
    "merge_double_checkdf.to_excel(os.path.join(path_to_main_output, f\"check_MethylDensity_and_Count_all_Samples.xlsx\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
